envoy/grpc/async_client.h

```diff
@@ -64,6 +65,11 @@ class RawAsyncStream {
    * limits
    */
   virtual bool isAboveWriteBufferHighWatermark() const PURE;
+
+  /**
+   * @returns the stream info object associated with this stream.
+   */
+  virtual const StreamInfo::StreamInfo& streamInfo() const PURE;
```

Comment:
I think this is probably the wrong interface for gRPC async client in Envoy. Since we have a dual implementation, we should have some class here that captures the stream info stats we want to export out, like bytes in/out and that in theory we can later populate with Google gRPC implementation stats from its API.

Reply:
Note that this is simply exporting whatever the http async client has. Even if we have a dual implementation, every consumer is once again forced to copy all the data into their internal data structures as this object is not guaranteed to be live when the access logger extracts the filter state. Is this dual implementation still worthwhile in this scenario given that consumers cant really store it?

Reply:
Another thing.. if the grpc async client alone exports a different stream info interface, then the implementations are now required to copy the http async client's stream info into this grpc async client new-stream info as part of every grpcStreamInfo() call. That feels like an unnecessary copy in the accessor method doesnt it?

Reply:
True, i guess it depends how much data we're talking about here - given the inherent overhead of a gRPC RPC probably a few bytes is noise, but if there is substantial string copying or large data structures I think so. What would be the size of the thing you are copying?

Reply:
Chatting offline, this data might be substantial protobuf structs like cluster metadata, I think reduced copy concerns are a fair consideration.

Reply:
we are talking about cluster metadata (protobuf structs) and few other things. These things tend to be of arbitrary size